### PR TITLE
bugfix(client): panic: send on closed channel

### PR DESCRIPTION
**Pull request by xaionaro**
Fix:

    panic: send on closed channel

    goroutine 751872 [running]:
    github.com/andreykaipov/goobs.(*Client).writeEvent(...)
        /home/xaionaro/.gvm/pkgsets/go1.22.1/global/pkg/mod/github.com/andreykaipov/goobs@v1.4.1/client.go:363
    github.com/andreykaipov/goobs.(*Client).handleOpcodes(0xc0020c81a0, 0xc0013846c0)
        /home/xaionaro/.gvm/pkgsets/go1.22.1/global/pkg/mod/github.com/andreykaipov/goobs@v1.4.1/client.go:338 +0x5a5
    created by github.com/andreykaipov/goobs.(*Client).connect in goroutine 751658
        /home/xaionaro/.gvm/pkgsets/go1.22.1/global/pkg/mod/github.com/andreykaipov/goobs@v1.4.1/client.go:200

Essentially by design we should close a channel only after we finished all possible writing to it. Thus moving the close statement of channel IncomingResponses to be called right after the writer to the channel is finished.

### CODE REVIEW

**Comment by andreykaipov** [comment]
Hi @xaionaro - can you rebase or pull in the latest main? I just wanna make sure the CI works with this change before merging this in!

Alternatively I think there's a checkbox somewhere that allows you to allow edits from maintainers and I can rebase for you.

**Comment by andreykaipov** [comment]
Hi @xaionaro - thanks for the PRs! Sorry it took a while. I didn't have time to figure out why the CI was failing in this repo. Now that's fixed in https://github.com/andreykaipov/goobs/pull/181, can you rebase this branch or pull in the latest changes from main? I wanna make sure it passes.

**Comment by andreykaipov** [comment]
Hi @xaionaro - thanks for the PRs! Sorry it took a while. I didn't have time to figure out why the CI was failing in this repo. But now that's fixed in https://github.com/andreykaipov/goobs/pull/181, can you rebase your PRs or pull in the latest changes from main to retrigger the CI? I wanna make sure that passes before merging them in! 🙏 

Alternatively I think there's a checkbox somewhere that allows you to allow edits from maintainers and I can rebase for you.

**Comment by xaionaro** [comment]
Wait, let me recheck one thing. I feel like I committed the wrong code.

**Comment by xaionaro** [comment]
@andreykaipov It seems like my fix was correct, but incomplete. Added a one more commit to fix some potential race conditions.

Let me also rebase my project on top of this commit and let's test this is prod for a week?

File: client.go
Comment (andreykaipov) [inline client.go:211]: `c.client.SendRequest` never writes to the `IncomingResponses` channel. It only reads from it:

https://github.com/andreykaipov/goobs/blob/060699aab72126c89d2af14ced42023659b8d51e/api/client.go#L106-L112

Comment (xaionaro) [reply client.go]: Indeed. Fixed the comment :)

File: client.go
Comment (andreykaipov) [inline client.go:221]: Sorry I'm having a hard time following the reasoning here. Previously `c.markDisconnected()` closed the opcodes channel first, at which point the loop in `c.handleOpcodes()` terminates, and since that's the only place where `c.client.IncomingResponses` is written to, that channel is closed thereafter:

https://github.com/andreykaipov/goobs/blob/bd2c39db013ff66b2c8703263c11268fa475d8aa/client.go#L141-L145

Can you help me understand what's different between this approach versus the above disconnection logic? I know things with concurrency are hard to reproduce, but a code snippet or something would help so I can try it myself to better understand the behaviour.

File: client.go
Comment (andreykaipov) [inline client.go:139]: I'm hesitant removing this bit because of situations where someone initiates a disconnect but the server might still have messages for the client:

https://github.com/andreykaipov/goobs/blob/314f152a96bd369e258239fcba015d185cd79977/client.go#L285-L294

Comment (xaionaro) [reply client.go:139]: The rationale is [ of course please let me know if I misunderstood the code :) ]:

From the perspective of the receiver code like this, it does not make any difference if it reads from the channel only once: https://goplay.tools/snippet/bt7F-8gxUGe . So this change would not hurt.

But there is a difference if they read it twice: https://goplay.tools/snippet/52vQxoRDPGY . So this change will fix the behavior in case channel `c.client.Disconnected:` is read multiple times.

File: client.go
Comment (xaionaro) [reply client.go:221]: > that channel is closed thereafter

If you mean that `close(c.client.IncomingResponses) ` is invoked after `close(c.client.Opcodes) ` then it is not good enough, because those are two separate goroutines (the one that closes the channels, and the other that consumes/writes (/to) the channels), so it is essentially a race condition.

> Can you help me understand what's different between this approach versus the above disconnection logic?

The old logic is:
* Start closing channels in one goroutine, while working with these channels in other goroutines. as a result we can do a lot of bad stuff, like writing to `IncomingResponses` after it is closed (because the writing routine checked `OpCodes` and it was not closed and proceeded to work with `IncomingResponses`, but in the meanwhile it was closed); and there are other examples.

The new logic is:
* Close `c.client.Disconnected` and `c.client.OpCodes` while locking `c.client.mutex` to avoid race condition in `SendRequest` (because `SendRequest` checks the `Disconnected` and proceeds with sending via `OpCodes` while holding `mutex`). The race condition here is now fixed.
* Close `c.client.IncomingResponses`  only after the writer (`handleOpcodes`) to this channel have finished. Thus race condition here is now also fixed.
* Close `c.client.IncomingResponses` only after `SendRequest` stopped reading from `c.client.IncomingResponses`, otherwise it will read a `nil`-valued `*opcodes.RequestResponse` and panic in line `if response.ID != id {`. And this order is achieved because `handleOpcodes` is finished only when `c.client.Opcodes` is closed, which happens only when `c.client.Disconnect` is closed, and thus `SendRequest` will not get into a branch where it reads from `c.client.IncomingResponses`, and no race condition here is guaranteed by the mutex (one more reason why we moved closure of some channels into function `Close` where we can lock `mutex`, which is a private field).

> I know things with concurrency are hard to reproduce, but a code snippet or something would help so I can try it myself to better understand the behaviour.

Here is an example: https://play.golang.com/p/4yy8Bm7IVDS
Please let me know if it is not descriptive enough :)

File: client.go
Comment (andreykaipov) [reply client.go:139]: This is something I didn't know, very cool!

File: client.go
Comment (andreykaipov) [reply client.go:221]: Sorry I didn't have time to check this out on the weekend. I think understand this better than I did before, but my knowledge is definitely lacking in there area so I really appreciate you walking me through it! Don't want to hold this up any longer. I hope this fixes https://github.com/andreykaipov/goobs/issues/182 :)

**Review by andreykaipov — APPROVED** [review]

**Pull request merged**
